parts/linux/cloud-init/artifacts/cse_main.sh

```diff
@@ -76,6 +77,20 @@ if [ "${ENABLE_SECURE_TLS_BOOTSTRAPPING}" = "true" ]; then
     logs_to_events "AKS.CSE.configureAndStartSecureTLSBootstrapping" configureAndStartSecureTLSBootstrapping
 fi
 
+# Node has been pre-provisioned before. Complete provisioning and skip the rest of the script.
+if [ -f /opt/azure/containers/preprovision.complete ]; then
```

Comment:
im not exactly sure what the SLA is with this new feature, though with your current implementation, any pre-provisioned VHD created from a VM in a particular cluster's node pool can only be used to create further nodes within that same cluster (and technically even the same node pool) - is that the goal?

this is because, from the looks of it, the pre-provisioned VHD will have a number of cluster and AP-specific data stored on disk which is used to provision the underlying node for k8s (cluster CA + apiserver certificate, bootstrap token, azure.json, to name a few)

Reply:
After we finish pre-provisioning we want to execute a user-provided script.
One of the main tasks of such script is to pull and cache user container images.

We want for the VM to be "almost functionable" but to not join the cluster. Then snapshot the VM and create a VHD in user space. Then this VHD will be used to spawn other VMs in the nodepool.

The goal is to have a customized VHD with user-specific software / images. So user can scale up quick. (Some users have customization that may take up to 1 hour to complete)